### DISCUSSION TITLE
This Serverless Function has timed out.

### DISCUSSION THREAD

**Discussion by racsou** [mitchmac/ServerlessWP#31] [Q&A]
This Serverless Function has timed out.

Your connection is working correctly.

Vercel is working correctly.

504: GATEWAY_TIMEOUT
Code: FUNCTION_INVOCATION_TIMEOUT
ID: cdg1::sr6rx-1695582672275-03be1404db59
![Capture](https://github.com/mitchmac/ServerlessWP/assets/120835349/31f830fa-d7a3-4835-8aa5-af4cc8d98486)

**Comment by racsou** [comment]
can you help me with this

**Comment by mitchmac** [comment]
Hi @racsou -- this will happen if the page takes too long to load and the serverless function times out. Can you share what URL/page you were trying to load so that we can figure out if there's a way around the timeout?

**Comment by Tobarrientos2** [comment]
Hi! I have this problem too

**Reply by Tobarrientos2** [reply]
When I refresh the page, the website is online, but when I go to /wp-admin, it doesn't remember my login username and password.

**Reply by mitchmac** [reply]
Hi! So you are getting the timeout when trying to login?

**Reply by Tobarrientos2** [reply]
Yes! However, when I try to refresh the page, I'm unable to sign in with my credentials because it's already set up with different ones.

**Reply by mitchmac** [reply]
Thanks for confirming. There are a bunch of factors, let's see if we can narrow it down. There might be clues in:

Do you have any WordPress plugins installed?
Where is the MySQL database hosted?
Is the site on Vercel, Netlify, or just AWS?

**Reply by Tobarrientos2** [reply]
So I did it on Vercel, and MySQL is hosted on PlanetScale.
I attempted to insert a user through the PlanetScale Terminal, and I was finally able to log in.
I'm not exactly sure what happened before, but there is a solution.

**Reply by juanmafernandezc** [reply]
Hi! How do you insert a user through the PlanetScale Terminal?

**Reply by Tobarrientos2** [reply]
You have to use SQL

**Comment by ibraidab** [comment]
If you are going to use vercel, modify in "api/index.js and add this line:

export const runtime = 'edge'; // 'nodejs' is the default

**Comment by braidata** [comment]
This is how the entire index.js should look like:

const serverlesswp = require('serverlesswp');

const { validate } = require('../util/install.js');
const { setup } = require('../util/directory.js');

export const runtime = 'edge'; // 'nodejs' is the default

// This is where all requests to WordPress are routed through. See vercel.json or netlify.toml for the redirection rules.
exports.handler = async function (event, context, callback) {
    // Move the /wp directory to /tmp/wp so that it is writeable.
    setup();

    // Send the request (event object) to the serverlesswp library. It includes the PHP server that allows WordPress to handle the request.
    let response = await serverlesswp({docRoot: '/tmp/wp', event: event});
    // Check to see if the database environment variables are in place.
    let checkInstall = validate(response);
    
    if (checkInstall) {
        return checkInstall;
    }
    else {
        // Return the response for serving.
        return response;
    }
}

**Reply by matheusrgg** [reply]
I did exaclty like this and steel not working

`const serverlesswp = require('serverlesswp');

const { validate } = require('../util/install.js');
const { setup } = require('../util/directory.js');
export const runtime = 'edge'; // 'nodejs' is the default
// This is where all requests to WordPress are routed through. See vercel.json or netlify.toml for the redirection rules.
exports.handler = async function (event, context, callback) {
    

    // Move the /wp directory to /tmp/wp so that it is writeable.
    setup();

    // Send the request (event object) to the serverlesswp library. It includes the PHP server that allows WordPress to handle the request.
    let response = await serverlesswp({docRoot: '/tmp/wp', event: event});
    // Check to see if the database environment variables are in place.
    let checkInstall = validate(response);
    
    if (checkInstall) {
        return checkInstall;
    }
    else {
        // Return the response for serving.
        return response;
    }
}`

**Reply by omegasoft7** [reply]
it doesnt work for me too

**Comment by YehonatanGitHub** [comment]
Hi any updates on this issue? I'm also not able to run successfully.

**Reply by CinicoDiogenes** [reply]
Nobody knows... :/ 

It's kind of frustrating. We just follow the tutorial, and bang! It is not working as it should, like in the video. 
I think that the solution is not to get the ''free pack''

**Reply by ibraidab** [reply]
Sí funciona, https://blog.braidata.com/ , no te frustres y seguí intentando. Sí es en vercel tenés que correrlo con edge.

**Reply by cisternascristian** [reply]
> Sí funciona, https://blog.braidata.com/ , no te frustres y seguí intentando. Sí es en vercel tenés que correrlo con edge.

En dónde tienes alojada la DB? A mi no me funcionó nunca.

**Reply by ibraidab** [reply]
La tenía en PlanetScale pero sacaron el Freemium y tuve que hacer el proceso de nuevo. Ahora la tengo en Aiven. Esta vez tuve que aumentar el tiempo de espera para que no me de time out. Lo que más demora es la instalación. Te conectas con prisma a la base de datos y vas a ver todas las tablas creadas.

**Reply by cisternascristian** [reply]
> La tenía en PlanetScale pero sacaron el Freemium y tuve que hacer el proceso de nuevo. Ahora la tengo en Aiven. Esta vez tuve que aumentar el tiempo de espera para que no me de time out. Lo que más demora es la instalación. Te conectas con prisma a la base de datos y vas a ver todas las tablas creadas.

También llegué a la solución de Aiven y le subí al tiempo de espera, se instaló pero me tira *"Error establishing a database connection"*, que también veo en tu blog... No sé si tiene que ver con Aiven y que requiere SSL u otra cosa, pero no he podido terminar con la configuración y echarlo a andar.

**Reply by ibraidab** [reply]
Lo que pasa es que el mio, lo dejé en pausa, el que reviví es este:

https://blog.inglesparatodosencasa.cl/

En septiembre tuve que reactivarlo, me costó mucho más que la primera, tuve que ser un poco más busquilla. Configuraste las variables de ambiente? en consola te da error CORS?

**Reply by cisternascristian** [reply]
Las variables de ambiente están bien, igual las volví a ingresar y le hice re-deploy, pero nada. En la consola me tira esto: 
> Failed to set up SSL because of the following SSL library error: SSL context is not usable without certificate and private key

**Reply by ibraidab** [reply]
Voy a revisar en el historial, porque creo que me enfrenté a ese error, reviso altiro.

**Comment by hoangngocthanh** [comment]
I got this error too
I hosted MySQL DB on AWS RDS and allowed public access, it is served on `us-est-1`
Is there any code enhancement for this error, I completely follow the tutorial.

**Comment by axjdroot** [comment]
Fixing the 504 Timeout Error on ServerlessWP
Hey everyone! This error almost always shows up during first-time setup. After going through this whole thread, it comes down to two situations — figure out which one you're in and follow the matching fix.

Which situation are you in?

Situation A → Your site loads, but the WordPress admin login doesn't work
Situation B → The site doesn't load at all


Situation A — Admin Login Not Working
You just need to create a fresh admin user directly in your database. Open your database console (PlanetScale terminal, Aiven console, phpMyAdmin — wherever your DB is hosted) and run these three queries one by one.
1. Create the new user:
sqlINSERT INTO wp_users 
  (user_login, user_pass, user_nicename, user_email, user_registered, user_status, display_name)
VALUES 
  ('newadmin', MD5('YourPasswordHere'), 'newadmin', 'you@email.com', NOW(), 0, 'New Admin');
2. Find the ID of the user you just created:
sqlSELECT ID FROM wp_users WHERE user_login = 'newadmin';
Note the number it returns — you'll need it next (let's say it's 5).
3. Give that user admin access (replace 5 with your actual number):
sqlINSERT INTO wp_usermeta (user_id, meta_key, meta_value)
VALUES
  (5, 'wp_capabilities', 'a:1:{s:13:"administrator";b:1;}'),
  (5, 'wp_user_level', '10');
Now go to /wp-admin and log in with newadmin and the password you set. Done! ✅

💡 Shortcut: I've attached a ready-made SQL file with onboarding already completed. Just import it into your database, then use the steps above to create your admin user — no need to go through the full setup.


Situation B — Site Not Loading At All
The serverless function is timing out before WordPress even finishes setting up. Two ways to fix it:
Fix 1 — Set up WordPress locally, then import (most reliable)

Install [LocalWP](https://localwp.com/) or [InstaWP](https://instawp.com/) on your computer and complete the WordPress setup there — no timeout limits on your own machine
Export your database as a .sql file
Import it into your remote database (Aiven, PlanetScale, AWS RDS, etc.)
Run this to update your site's URL in the database:

sqlUPDATE wp_options SET option_value = 'https://your-vercel-url.vercel.app' WHERE option_name = 'siteurl';
UPDATE wp_options SET option_value = 'https://your-vercel-url.vercel.app' WHERE option_name = 'home';

Deploy to Vercel — it'll skip the install since the database is already set up ✅

Fix 2 — Switch to Edge Runtime in Vercel (quick code change)
Open api/index.js in your project and add one line near the top:
jsconst serverlesswp = require('serverlesswp');
const { validate } = require('../util/install.js');
const { setup } = require('../util/directory.js');

export const runtime = 'edge'; // 👈 Add this line

exports.handler = async function (event, context, callback) {
  setup();
  let response = await serverlesswp({ docRoot: '/tmp/wp', event: event });
  let checkInstall = validate(response);
  if (checkInstall) {
    return checkInstall;
  } else {
    return response;
  }
};

⚠️ Still not working after this? The problem is likely your database connection, not the timeout. Double-check your Vercel environment variables: WORDPRESS_DB_HOST, WORDPRESS_DB_USER, WORDPRESS_DB_PASSWORD, WORDPRESS_DB_NAME.


SSL Error? (Aiven / AWS RDS users)
If you see something like SSL context is not usable without certificate and private key or Error establishing a database connection, your database is requiring SSL but it's not configured. Check your database provider's dashboard for SSL connection details and make sure they're reflected in your Vercel environment variables.

How to avoid this error entirely

Always complete WordPress setup locally first, then migrate the database — this is the single most reliable approach
Make sure your database is hosted in the same region as your Vercel deployment. A database in us-east-1 connecting to a Vercel function in Paris will be slow enough to cause timeouts
Add export const runtime = 'edge' to api/index.js before your first deploy
Set all your environment variables in Vercel before deploying


Reply with your database provider and Vercel region if you're still stuck and I'll help narrow it down! 🙌
